proposed/http-middleware/middleware.md

```diff
@@ -2,102 +2,203 @@ HTTP Middleware
 ===============
 
 This document describes a common standard for HTTP middleware components using
-[PSR-7][psr7] compliant HTTP messages.
+HTTP Messages defined by [PSR-7](http://www.php-fig.org/psr/psr-7/).
 
 HTTP middleware has been an important concept on other web development platforms
-for a good number of years, and since the inception of PSR-7 has been growing
-increasingly popular in the PHP world.
+for a good number of years, and since the introduction of a formal HTTP Messages
+standard has been growing increasingly popular with web frameworks.
 
-An informal (de-facto) standard for the signature of middleware components is
-already widely in use, and this PSR serves to provide a formal definition in
-the form of a common standard interface for middleware components.
-
-The interface described in this document is an abstraction of an individual
-middleware component.
+The interfaces described in this document are abstractions for HTTP middleware
+and the containers that are used to process HTTP requests.
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
 "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
-interpreted as described in [RFC 2119][rfc2119].
+interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119).
+
+### References
 
-[psr7]: http://www.php-fig.org/psr/psr-7/
-[rfc2119]: http://tools.ietf.org/html/rfc2119
+- [RFC 2119](http://tools.ietf.org/html/rfc2119)
+- [PSR-7](http://www.php-fig.org/psr/psr-7/)
 
 ## 1. Specification
 
 An HTTP middleware component is an individual component participating, together
 with other middleware components, in the processing of an incoming HTTP Request
 and the creation of a resulting HTTP Response, as defined by PSR-7.
 
-Middleware components MUST implement this interface.
+Middleware consumers such as frameworks and middleware stack containers MUST
+type hint against one of the middleware interfaces:
```

Comment:
_type hint_ should be _type declaration_. These are not hints, they're enforced. `/** @var Type */ is a hint, but these are declarations. See also http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.